tools/scylla-sstable.cc

```diff
@@ -3032,6 +3053,273 @@ for tests.
+R"(
+Read the sstable(s) and pass the resulting fragment stream (see the help for the
+validation operation for mode details on that) to the script specified by
+--script-file. The content is traversed over in its natural order.
+Currently only Lua scripts are supported.
+
+# The Lua API
+
+The fragment stream is fed to the script via the following methods defined by the
+script:
+
+## on_start_of_stream()
+
+Called on the very start of the stream. Can be used to initialize global state.
+
+## on_new_sstable(sst)
+
+Called on the start of each sstable. Parameter is of type sstable. When sstables
+are merged (--merge) the parameter is nil.
+Returns whether to stop. If true, on_end_of_sstable() is called, skipping the
+content of the sstable (or that of the entire stream if --merge is used).
+If false, consumption follows with the content of the sstable.
+
+## consume_partition_start(ps)
+
+Called on the start of each partition. Parameter is of type partition_start.
+Returns whether to stop. If true, consume_partition_end() is called, skipping
+the content of the partition. If false, consumption follows with the content of
+the partition.
+
+## consume_static_row(sr)
+
+Called if the partition has a static row. Parameter is of type static_row.
+Returns whether to stop. If true, consume_partition_end() is called, skipping
+the remaining content of the partition. If false, consumption follows with the
+remaining content of the partition.
+
+## consume_clustering_row(cr)
+
+Called for each clustering row. Parameter is of type clustering_row.
+Returns whether to stop. If true, consume_partition_end() is called, skipping
+the remaining content of the partition. If false, consumption follows with the
+remaining content of the partition.
+
+## consume_range_tombstone_change(crt)
+
+Called for each range tombstone change. Parameter is of type
+range_tombstone_change.
+Returns whether to stop. If true, consume_partition_end() is called, skipping
+the remaining content of the partition. If false, consumption follows with the
+remaining content of the partition.
+
+## consume_partition_end()
+
+Called at the end of the partition.
+Returns whether to stop. If true, on_end_of_sstable() is called, skipping
+the remaining content of the sstable. If false, consumption follows with the
+remaining content of the sstable.
+
+## on_end_of_sstable()
+
+Called at the end of the sstable.
+Returns whether to stop. If true, on_end_of_stream() is called, skipping
+the remaining content of the stream. If false, consumption follows with the
+remaining content of the stream.
+
+## on_end_of_stream()
+
+Called at the very end of the stream.
+
+Note that the script doesn't have to define all the above methods. It can define
+those that it is interested in and omit the rest. For example a script only
+interested in partitions, can define only consume_partition_start() and nothing
+else. Therefore a completely empty script is also valid, although not very
+useful.
```

Comment:
How does one choose not to write the default json output?

Reply:
At first I didn't understand this comment at all, but then I saw that you thought this scripting is similar to `awk`. Actually, it is very different, the script actually implementing an sstable consumer interface. Printing happens via explicit action by the script.

Reply:
Yes. I think we should make it easy to print out stuff for the filtering use case. But there's no need to do everything all at once.